File: ActionUINodeJS/test_native_api_surface.js
'use strict';
/**
 * test_native_api.js — Native module API surface smoke tests.
 *
 * Mirrors test_api_surface.py from ActionUIPython.
 * Tests the N-API binding layer directly without a running NSApplication.
 *
 * Run: node test_native_api.js
 */

const native = require('./index.js')._native;

let failures = [];

function check(label, condition) {
    if (condition) {
        console.log(`  [PASS] ${label}`);
    } else {
        console.error(`  [FAIL] ${label}`);
        failures.push(label);
    }
    return condition;
}

function testModuleConstants() {
    console.log('\n=== Module constants ===');
    check('LOG_ERROR === 1',   native.LOG_ERROR === 1);
    check('LOG_WARNING === 2', native.LOG_WARNING === 2);
    check('LOG_INFO === 3',    native.LOG_INFO === 3);
    check('LOG_DEBUG === 4',    native.LOG_DEBUG === 4);
    check('LOG_VERBOSE === 5', native.LOG_VERBOSE === 5);
}

function testVersion() {
    console.log('\n=== Version ===');
    const v = native.getVersion();
    check('getVersion() returns a string', typeof v === 'string' && v.length > 0);
}

function testLogging() {
    console.log('\n=== Logging ===');
    try {
        native.setLogger((msg, level) => {});
        native.log('test message', native.LOG_INFO);
        native.setLogger(null);
        check('setLogger/log/clear works without throw', true);
    } catch (e) {
        check('setLogger/log/clear works without throw', false);
    }
}

function testErrorLifecycle() {
    console.log('\n=== Error lifecycle ===');
    native.clearError();
    check('clearError() does not throw', true);
    const err = native.getLastError();
    check('getLastError() returns null after clear', err === null);
}

function testActionHandlerLifecycle() {
    console.log('\n=== Action handler lifecycle ===');
    const handler = (aid, wid, vid, vpid, ctx) => {};
    try {
        native.registerActionHandler('test.action', handler);
        native.unregisterActionHandler('test.action');
        native.setDefaultActionHandler(handler);
        native.setDefaultActionHandler(null);
        check('register/unregister/setDefault works without throw', true);
    } catch (e) {
        check('register/unregister/setDefault works without throw', false);
    }
}

function testAppCalls() {
    console.log('\n=== App lifecycle calls ===');
    try {
        native.appLoadMenuBar();
        native.appLoadMenuBar('[]');
        native.appSetWillFinishLaunching(() => {});
        native.appSetDidFinishLaunching(() => {});
        native.appSetWillBecomeActive(() => {});
        native.appSetDidBecomeActive(() => {});
        native.appSetWillResignActive(() => {});
        native.appSetDidResignActive(() => {});
        native.appSetWillTerminate(() => {});
        native.appSetShouldTerminate(() => true);
        native.appSetWindowWillClose(() => {});
        native.appSetWindowWillPresent(() => {});
        native.appSetWillFinishLaunching(null);
        native.appSetDidFinishLaunching(null);
        native.appSetWillBecomeActive(null);
        native.appSetDidBecomeActive(null);
        native.appSetWillResignActive(null);
        native.appSetDidResignActive(null);
        native.appSetWillTerminate(null);
        native.appSetShouldTerminate(null);
        native.appSetWindowWillClose(null);
        native.appSetWindowWillPresent(null);
        check('All app lifecycle setters work without throw', true);
    } catch (e) {
        check('All app lifecycle setters work without throw', false);
    }
}

function testTypeSpecificSettersAndGetters(uuid) {
    console.log('\n=== Type-specific setters and getters ===');
    const vid = 1;
    const vpid = 0;

    let r = native.setIntValue(uuid, vid, vpid, 42);
    check('setIntValue returns true', r === true);
    r = native.setIntValue(uuid, vid, vpid, -7);
    check('setIntValue(-7) returns true', r === true);
    r = native.setDoubleValue(uuid, vid, vpid, 3.14);
    check('setDoubleValue returns true', r === true);
    r = native.setBoolValue(uuid, vid, vpid, true);
    check('setBoolValue(true) returns true', r === true);
    r = native.setBoolValue(uuid, vid, vpid, false);
    check('setBoolValue(false) returns true', r === true);
    r = native.setStringValue(uuid, vid, vpid, 'hello world');
    check('setStringValue returns true', r === true);
    r = native.setStringValue(uuid, vid, vpid, '');
    check('setStringValue(empty) returns true', r === true);

    check('getIntValue returns null (no UI)', native.getIntValue(uuid, vid, vpid) === null);
    check('getDoubleValue returns null', native.getDoubleValue(uuid, vid, vpid) === null);
    check('getBoolValue returns null', native.getBoolValue(uuid, vid, vpid) === null);
    check('getStringValue returns null', native.getStringValue(uuid, vid, vpid) === null);
}

function testViewPartIdIsolation(uuid) {
    console.log('\n=== View part ID isolation ===');
    const vid = 2;
    for (const vpid of [0, 1, 2]) {
        const r = native.setIntValue(uuid, vid, vpid, 100 + vpid);
        check(`setIntValue(vpid=${vpid}) returns true`, r === true);
    }
}

function testDefaultViewPartId(uuid) {
    console.log('\n=== Default view part ID ===');
    const vid = 3;
    const r = native.setIntValue(uuid, vid, 0, 999);
    check('setIntValue(vpid=0) returns true', r === true);
}

function testStringBoundaryCases(uuid) {
    console.log('\n=== String boundary cases ===');
    const vid = 4;
    const vpid = 0;
    for (const s of ['', 'a', 'x'.repeat(1000)]) {
        const r = native.setStringValue(uuid, vid, vpid, s);
        check(`setStringValue("${s.slice(0, 10)}...") returns true`, r === true);
    }
}

function testValueFromString(uuid) {
    console.log('\n=== Value from string ===');
    const vid = 5;
    const vpid = 0;

    let r = native.setValueFromString(uuid, vid, vpid, 'plain text');
    check('setValueFromString(plain) returns true', r === true);
    r = native.setValueFromString(uuid, vid, vpid, '**bold**', 'markdown');
    check('setValueFromString(markdown) returns true', r === true);
    r = native.setValueFromString(uuid, vid, vpid, 'another value');
    check('setValueFromString default format returns true', r === true);
}

function testValueFromJson(uuid) {
    console.log('\n=== Value from JSON ===');
    const vid = 6;
    const vpid = 0;

    let r = native.setValueFromJSON(uuid, vid, vpid, JSON.stringify({ k: 'v' }));
    check('setValueFromJSON(object) returns true', r === true);
    r = native.setValueFromJSON(uuid, vid, vpid, '{}');
    check('setValueFromJSON({}) returns true', r === true);
    r = native.setValueFromJSON(uuid, vid, vpid, '[]');
    check('setValueFromJSON([]) returns true', r === true);
}

function testElementColumnCount(uuid) {
    console.log('\n=== Element column count ===');
    const vid = 10;
    const r = native.getElementColumnCount(uuid, vid);
    check('getElementColumnCount returns an integer', typeof r === 'number');
}

function testElementRows(uuid) {
    console.log('\n=== Element rows ===');
    const vid = 11;
    const rows = [['A', 'B'], ['C', 'D']];
    let r = native.setElementRowsJSON(uuid, vid, JSON.stringify(rows));
    check('setElementRowsJSON returns true', r === true);
    r = native.appendElementRowsJSON(uuid, vid, JSON.stringify([['E', 'F']]));
    check('appendElementRowsJSON returns true', r === true);
    r = native.clearElementRows(uuid, vid);
    check('clearElementRows does not throw', r === true || r === undefined);
}

function testElementProperty(uuid) {
    console.log('\n=== Element property ===');
    const vid = 12;
    const r = native.setElementPropertyJSON(uuid, vid, 'hidden', JSON.stringify(true));
    check('setElementPropertyJSON returns true', r === true);
}

function testElementState(uuid) {
    console.log('\n=== Element state ===');
    const vid = 13;
    const key = 'counter';
    let r = native.setElementStateJSON(uuid, vid, key, JSON.stringify(0));
    check('setElementStateJSON returns true', r === true);
    r = native.setElementStateFromString(uuid, vid, key, '42');
    check('setElementStateFromString returns true', r === true);
}

function testElementInfo(uuid) {
    console.log('\n=== Element info ===');
    const v = native.getElementInfoJSON(uuid);
    check('getElementInfoJSON returns null or parses', v === null || typeof v === 'string');
}

function testNoopCalls(uuid) {
    console.log('\n=== No-op calls (getters without UI) ===');
    const vid = 99;
    const vpid = 0;

    native.setIntValue(uuid, vid, vpid, 1);
    native.setDoubleValue(uuid, vid, vpid, 1.0);
    native.setBoolValue(uuid, vid, vpid, true);
    native.setStringValue(uuid, vid, vpid, 'x');
    native.setValueFromString(uuid, vid, vpid, 'x');
    native.setValueFromJSON(uuid, vid, vpid, '{}');
    native.setElementRowsJSON(uuid, vid, '[]');
    native.appendElementRowsJSON(uuid, vid, '[]');
    native.clearElementRows(uuid, vid);
    native.setElementPropertyJSON(uuid, vid, 'x', '{}');
    native.setElementStateJSON(uuid, vid, 'x', '{}');
    native.setElementStateFromString(uuid, vid, 'x', 'x');

    check('getIntValue returns null', native.getIntValue(uuid, vid, vpid) === null);
    check('getDoubleValue returns null', native.getDoubleValue(uuid, vid, vpid) === null);
    check('getBoolValue returns null', native.getBoolValue(uuid, vid, vpid) === null);
    check('getStringValue returns null', native.getStringValue(uuid, vid, vpid) === null);
    check('getValueAsString returns null', native.getValueAsString(uuid, vid, vpid) === null);
    check('getValueAsJSON returns null', native.getValueAsJSON(uuid, vid, vpid) === null);
    check('getElementRowsJSON returns null', native.getElementRowsJSON(uuid, vid) === null);
    check('getElementColumnCount returns 0', native.getElementColumnCount(uuid, vid) === 0);
    check('getElementPropertyJSON returns null', native.getElementPropertyJSON(uuid, vid, 'x') === null);
    check('getElementStateJSON returns null', native.getElementStateJSON(uuid, vid, 'x') === null);
    check('getElementStateString returns null', native.getElementStateString(uuid, vid, 'x') === null);
}

function testModalNoop(uuid) {
    console.log('\n=== Modal no-op calls ===');
    const calls = [
        () => native.presentModal(uuid, '{}', 'json', 'sheet', null),
        () => native.dismissModal(uuid),
        () => native.presentAlert(uuid, 'title', 'message', null),
        () => native.presentConfirmationDialog(uuid, 'title', 'message', '[]'),
        () => native.dismissDialog(uuid),
    ];
    for (const fn of calls) {
        try { fn(); } catch {}
    }
    check('Modal calls do not crash', true);
}

function testLoadHostingController(uuid) {
    console.log('\n=== Load hosting controller ===');
    let threw = false;
    try {
        native.loadHostingController('file:///nonexistent/path.json', uuid, true);
    } catch (e) {
        threw = true;
    }
    check('loadHostingController returns without crash or throws on missing file', true);
}

function testMenuBarNoop() {
    console.log('\n=== Menu bar no-op ===');
    try {
        native.appLoadMenuBar('not valid json');
        check('appLoadMenuBar(invalid JSON) does not crash', true);
    } catch (e) {
        check('appLoadMenuBar(invalid JSON) does not crash', false);
    }
}

function testAppControl() {
    console.log('\n=== App control ===');
    const calls = [
        () => native.appSetName('TestApp'),
        () => native.appSetIcon(null),
    ];
    for (const fn of calls) {
        try { fn(); } catch {}
    }
    check('appSetName/appSetIcon do not throw', true);
}

function testFilePanels() {
    console.log('\n=== File panel API surface ===');
    check('appRunOpenPanel is a function', typeof native.appRunOpenPanel === 'function');
    check('appRunSavePanel is a function', typeof native.appRunSavePanel === 'function');
    check('appRunAlert is a function', typeof native.appRunAlert === 'function');
}

function testTypeChecking() {
    console.log('\n=== Type checking (non-function must throw TypeError) ===');
    const setters = [
        ['appSetWillFinishLaunching', native.appSetWillFinishLaunching.bind(native)],
        ['appSetShouldTerminate', native.appSetShouldTerminate.bind(native)],
        ['appSetWindowWillClose', native.appSetWindowWillClose.bind(native)],
        ['setLogger', native.setLogger.bind(native)],
    ];

    for (const [name, setter] of setters) {
        let threw = false;
        try { setter(42); } catch { threw = true; }
        check(`native.${name}(42) throws`, threw);
    }
}

function testNullClearing() {
    console.log('\n=== Null handler clearing ===');
    try {
        native.appSetWillFinishLaunching(() => {});
        native.appSetWillFinishLaunching(null);
        check('appSetWillFinishLaunching(null) clears without throw', true);
    } catch (e) {
        check('appSetWillFinishLaunching(null) clears without throw', false);
    }

    try {
        native.appSetShouldTerminate(() => true);
        native.appSetShouldTerminate(null);
        check('appSetShouldTerminate(null) clears without throw', true);
    } catch (e) {
        check('appSetShouldTerminate(null) clears without throw', false);
    }

    try {
        native.appSetWindowWillClose(() => {});
        native.appSetWindowWillClose(null);
        check('appSetWindowWillClose(null) clears without throw', true);
    } catch (e) {
        check('appSetWindowWillClose(null) clears without throw', false);
    }
}

function main() {
    console.log('ActionUI Native Module — API Surface Test');
    console.log('='.repeat(55));
    console.log('(No NSApplication run loop — tests the N-API binding only)');

    testModuleConstants();
    testVersion();
    testLogging();
    testErrorLifecycle();
    testActionHandlerLifecycle();
    testAppCalls();

    const testUuid = require('crypto').randomUUID();

    testTypeSpecificSettersAndGetters(testUuid);
    testViewPartIdIsolation(testUuid);
    testDefaultViewPartId(testUuid);
    testStringBoundaryCases(testUuid);
    testValueFromString(testUuid);
    testValueFromJson(testUuid);
    testElementColumnCount(testUuid);
    testElementRows(testUuid);
    testElementProperty(testUuid);
    testElementState(testUuid);
    testElementInfo(testUuid);
    testNoopCalls(testUuid);
    testModalNoop(testUuid);
    testLoadHostingController(testUuid);
    testMenuBarNoop();
    testAppControl();
    testFilePanels();
    testTypeChecking();
    testNullClearing();

    console.log('\n' + '='.repeat(55));
    if (failures.length === 0) {
        console.log('All API surface tests PASSED.');
    } else {
        console.error(`FAILED — ${failures.length} check(s):`);
        for (const f of failures) console.error(`  - ${f}`);
        process.exit(1);
    }
}

main();

File: ActionUIPython/test_native_api_surface.py
#!/usr/bin/env python3
"""
Comprehensive API surface test for the _actionui C extension.

Exercises every function in the module table to catch signature mismatches,
segfaults, and argument-order bugs without needing a live UI or NSApp run loop.

Run with: python3 test_native_api.py
"""

import _actionui
import json
import uuid


def test_module_constants():
    assert _actionui.LOG_ERROR == 1
    assert _actionui.LOG_WARNING == 2
    assert _actionui.LOG_INFO == 3
    assert _actionui.LOG_DEBUG == 4
    assert _actionui.LOG_VERBOSE == 5
    print("  module_constants: OK")


def test_version():
    v = _actionui.get_version()
    assert v is not None and isinstance(v, str) and len(v) > 0
    print(f"  get_version: '{v}'")


def test_logging():
    _actionui.set_logger(lambda msg, level: None)
    _actionui.log("test message", 2)
    _actionui.set_logger(None)
    print("  logging: OK")


def test_error_lifecycle():
    _actionui.clear_error()
    assert _actionui.get_last_error() is None
    print("  error_lifecycle: OK")


def test_action_handler_lifecycle():
    def handler(aid, wid, vid, vpid, ctx):
        pass
    _actionui.register_action_handler("test.action", handler)
    _actionui.unregister_action_handler("test.action")
    _actionui.set_default_action_handler(handler)
    _actionui.set_default_action_handler(None)
    print("  action_handler_lifecycle: OK")


def test_app_calls():
    _actionui.app_load_menu_bar()
    _actionui.app_load_menu_bar("[]")
    _actionui.app_set_will_finish_launching(lambda: None)
    _actionui.app_set_did_finish_launching(lambda: None)
    _actionui.app_set_will_become_active(lambda: None)
    _actionui.app_set_did_become_active(lambda: None)
    _actionui.app_set_will_resign_active(lambda: None)
    _actionui.app_set_did_resign_active(lambda: None)
    _actionui.app_set_will_terminate(lambda: None)
    _actionui.app_set_should_terminate(lambda: True)
    _actionui.app_set_window_will_close(lambda w: None)
    _actionui.app_set_window_will_present(lambda w: None)
    _actionui.app_set_will_finish_launching(None)
    _actionui.app_set_did_finish_launching(None)
    _actionui.app_set_will_become_active(None)
    _actionui.app_set_did_become_active(None)
    _actionui.app_set_will_resign_active(None)
    _actionui.app_set_did_resign_active(None)
    _actionui.app_set_will_terminate(None)
    _actionui.app_set_should_terminate(None)
    _actionui.app_set_window_will_close(None)
    _actionui.app_set_window_will_present(None)
    print("  app_calls: OK")


def test_type_specific_setters_and_getters(uuid_str):
    vid = 1
    vpid = 0

    r = _actionui.set_int_value(uuid_str, vid, vpid, 42)
    assert r is True, f"set_int_value returned {r}"
    _actionui.set_int_value(uuid_str, vid, vpid, -7)
    r = _actionui.set_double_value(uuid_str, vid, vpid, 3.14)
    assert r is True
    r = _actionui.set_bool_value(uuid_str, vid, vpid, True)
    assert r is True
    r = _actionui.set_bool_value(uuid_str, vid, vpid, False)
    assert r is True
    r = _actionui.set_string_value(uuid_str, vid, vpid, "hello world")
    assert r is True
    r = _actionui.set_string_value(uuid_str, vid, vpid, "")
    assert r is True

    # Getters return None without a loaded UI.
    assert _actionui.get_int_value(uuid_str, vid, vpid) is None
    assert _actionui.get_double_value(uuid_str, vid, vpid) is None
    assert _actionui.get_bool_value(uuid_str, vid, vpid) is None
    assert _actionui.get_string_value(uuid_str, vid, vpid) is None
    print("  type_specific_setters_and_getters: OK")


def test_view_part_id_isolation(uuid_str):
    vid = 2
    for vpid in (0, 1, 2):
        r = _actionui.set_int_value(uuid_str, vid, vpid, 100 + vpid)
        assert r is True
    print("  view_part_id_isolation: OK")


def test_default_view_part_id(uuid_str):
    vid = 3
    r = _actionui.set_int_value(uuid_str, vid, 0, 999)
    assert r is True
    print("  default_view_part_id: OK")


def test_string_boundary_cases(uuid_str):
    vid = 4
    vpid = 0
    for s in ("", "a", "x" * 1000):
        r = _actionui.set_string_value(uuid_str, vid, vpid, s)
        assert r is True
    print("  string_boundary_cases: OK")


def test_value_from_string(uuid_str):
    vid = 5
    vpid = 0

    r = _actionui.set_value_from_string(uuid_str, vid, vpid, "plain text")
    assert r is True
    r = _actionui.set_value_from_string(uuid_str, vid, vpid, "**bold**", "markdown")
    assert r is True
    r = _actionui.set_value_from_string(uuid_str, vid, vpid, "another value")
    assert r is True
    print("  set_value_from_string: OK")


def test_value_from_json(uuid_str):
    vid = 6
    vpid = 0

    r = _actionui.set_value_from_json(uuid_str, vid, vpid, json.dumps({"k": "v"}))
    assert r is True
    r = _actionui.set_value_from_json(uuid_str, vid, vpid, "{}")
    assert r is True
    r = _actionui.set_value_from_json(uuid_str, vid, vpid, "[]")
    assert r is True
    print("  value_from_json: OK")


def test_element_column_count(uuid_str):
    vid = 10
    r = _actionui.get_element_column_count(uuid_str, vid)
    assert isinstance(r, int)
    print("  element_column_count: OK")


def test_element_rows(uuid_str):
    vid = 11
    rows = [["A", "B"], ["C", "D"]]
    r = _actionui.set_element_rows_json(uuid_str, vid, json.dumps(rows))
    assert r is True
    r = _actionui.append_element_rows_json(uuid_str, vid, json.dumps([["E", "F"]]))
    assert r is True
    _actionui.clear_element_rows(uuid_str, vid)
    print("  element_rows: OK")


def test_element_property(uuid_str):
    vid = 12
    r = _actionui.set_element_property_json(uuid_str, vid, "hidden", json.dumps(True))
    assert r is True
    print("  element_property: OK")


def test_element_state(uuid_str):
    vid = 13
    key = "counter"
    r = _actionui.set_element_state_json(uuid_str, vid, key, json.dumps(0))
    assert r is True
    r = _actionui.set_element_state_from_string(uuid_str, vid, key, "42")
    assert r is True
    print("  element_state: OK")


def test_element_info(uuid_str):
    v = _actionui.get_element_info_json(uuid_str)
    assert v is None or isinstance(json.loads(v), dict)
    print("  element_info: OK")


def test_noop_calls(uuid_str):
    vid = 99
    vpid = 0
    _actionui.set_int_value(uuid_str, vid, vpid, 1)
    _actionui.set_double_value(uuid_str, vid, vpid, 1.0)
    _actionui.set_bool_value(uuid_str, vid, vpid, True)
    _actionui.set_string_value(uuid_str, vid, vpid, "x")
    _actionui.set_value_from_string(uuid_str, vid, vpid, "x")
    _actionui.set_value_from_json(uuid_str, vid, vpid, "{}")
    _actionui.set_element_rows_json(uuid_str, vid, "[]")
    _actionui.append_element_rows_json(uuid_str, vid, "[]")
    _actionui.clear_element_rows(uuid_str, vid)
    _actionui.set_element_property_json(uuid_str, vid, "x", "{}")
    _actionui.set_element_state_json(uuid_str, vid, "x", "{}")
    _actionui.set_element_state_from_string(uuid_str, vid, "x", "x")
    assert _actionui.get_int_value(uuid_str, vid, vpid) is None
    assert _actionui.get_double_value(uuid_str, vid, vpid) is None
    assert _actionui.get_bool_value(uuid_str, vid, vpid) is None
    assert _actionui.get_string_value(uuid_str, vid, vpid) is None
    assert _actionui.get_value_as_string(uuid_str, vid, vpid) is None
    assert _actionui.get_value_as_json(uuid_str, vid, vpid) is None
    assert _actionui.get_element_rows_json(uuid_str, vid) is None
    assert _actionui.get_element_column_count(uuid_str, vid) == 0
    assert _actionui.get_element_property_json(uuid_str, vid, "x") is None
    assert _actionui.get_element_state_json(uuid_str, vid, "x") is None
    assert _actionui.get_element_state_string(uuid_str, vid, "x") is None
    print("  noop_calls: OK")


def test_modal_noop(uuid_str):
    for fn in [
        lambda: _actionui.present_modal(uuid_str, "{}", "json", "sheet", None),
        lambda: _actionui.dismiss_modal(uuid_str),
        lambda: _actionui.present_alert(uuid_str, "title", "message", None),
        lambda: _actionui.present_confirmation_dialog(uuid_str, "title", "message", "[]"),
        lambda: _actionui.dismiss_dialog(uuid_str),
    ]:
        try:
            fn()
        except RuntimeError:
            pass
    print("  modal_noop: OK")


def test_load_hosting_controller(uuid_str):
    ptr = _actionui.load_hosting_controller("file:///nonexistent/path.json", uuid_str, True)
    assert ptr is not None
    print("  load_hosting_controller: OK")


def main():
    print("ActionUI _actionui C Extension — API Surface Test")
    print("=" * 55)

    test_module_constants()
    test_version()
    test_logging()
    test_error_lifecycle()
    test_action_handler_lifecycle()
    test_app_calls()

    test_uuid = str(uuid.uuid4())

    test_type_specific_setters_and_getters(test_uuid)
    test_view_part_id_isolation(test_uuid)
    test_default_view_part_id(test_uuid)
    test_string_boundary_cases(test_uuid)
    test_value_from_string(test_uuid)
    test_value_from_json(test_uuid)
    test_element_column_count(test_uuid)
    test_element_rows(test_uuid)
    test_element_property(test_uuid)
    test_element_state(test_uuid)
    test_element_info(test_uuid)
    test_noop_calls(test_uuid)
    test_modal_noop(test_uuid)
    test_load_hosting_controller(test_uuid)

    print()
    print("All API surface tests passed.")


if __name__ == "__main__":
    main()